File: build_workbook.mjs
#!/usr/bin/env node
import fs from "node:fs/promises";
import path from "node:path";
import { SpreadsheetFile, Workbook } from "@oai/artifact-tool";

const [jsonPathArg] = process.argv.slice(2);
if (!jsonPathArg) {
  console.error("Usage: node tax_workflow_tool/build_workbook.mjs <workflow_result.json>");
  process.exit(1);
}

const jsonPath = path.resolve(jsonPathArg);
const outDir = path.dirname(jsonPath);
const data = JSON.parse(await fs.readFile(jsonPath, "utf8"));

const workbook = Workbook.create();
const finalSheet = workbook.worksheets.add("最终输出");
const workflowSheet = workbook.worksheets.add("筛选口径");
const currencySheet = workbook.worksheets.add("币种汇总");
const detailSheet = workbook.worksheets.add("命中明细");

function writeTable(sheet, startCell, headers, rows) {
  const values = [headers, ...rows.map((row) => headers.map((header) => row[header] ?? ""))];
  const range = sheet.getRange(startCell).resize(values.length, headers.length);
  range.values = values;
  range.format = { wrapText: true };
  sheet.getRange(startCell).resize(1, headers.length).format = {
    fill: "#1F4E78",
    font: { bold: true, color: "#FFFFFF" },
    wrapText: true,
  };
  range.format.autofitColumns();
}

finalSheet.showGridLines = false;
finalSheet.getRange("A1:E1").merge();
finalSheet.getRange("A1").values = [["A1-A3未代扣代缴工作流输出"]];
finalSheet.getRange("A1").format = {
  fill: "#17365D",
  font: { bold: true, color: "#FFFFFF", size: 16 },
};
finalSheet.getRange("A3:B8").values = [
  ["算税月", data.activity_period],
  ["算税国家", data.tax_country],
  ["源文件", data.source_file],
  ["已注册税号国家", data.registered_vat_countries.join(", ")],
  ["未注册欧盟国家", data.unregistered_eu_countries.join(", ")],
  ["当月汇率", Object.entries(data.rates_for_month).map(([k, v]) => `${k}:${v ?? "未填"}`).join(", ")],
];
finalSheet.getRange("A3:A8").format = { fill: "#D9EAF7", font: { bold: true } };
finalSheet.getRange("B3:B8").format = { wrapText: true };
writeTable(finalSheet, "A10", ["A项", "名称", "命中行数", "EUR金额总和", "缺少汇率行数"], data.summary);
writeTable(finalSheet, "A16", ["项目", "公式", "金额EUR"], data.final);
finalSheet.getRange("D11:D13").format.numberFormat = "#,##0.00";
finalSheet.getRange("C17:C18").format.numberFormat = "#,##0.00";
finalSheet.freezePanes.freezeRows(10);

writeTable(workflowSheet, "A1", ["A项", "名称", "筛选口径", "金额列"], data.summary);
workflowSheet.freezePanes.freezeRows(1);

writeTable(currencySheet, "A1", ["A项", "币种", "行数", "原币合计", "折EUR金额", "缺少汇率行数"], data.currency);
currencySheet.getRange("D2:E200").format.numberFormat = "#,##0.00";
currencySheet.freezePanes.freezeRows(1);

const detailHeaders = [
  "A项",
  "A项名称",
  "ACTIVITY_PERIOD",
  "TAX_COLLECTION_RESPONSIBILITY",
  "TRANSACTION_TYPE",
  "EXPORT_OUTSIDE_EU",
  "PRICE_OF_ITEMS_VAT_RATE_PERCENT",
  "SALE_DEPART_COUNTRY",
  "SALE_ARRIVAL_COUNTRY",
  "BUYER_VAT_NUMBER",
  "TRANSACTION_CURRENCY_CODE",
  "TOTAL_ACTIVITY_VALUE_AMT_VAT_INCL",
  "_original_amount",
  "_eur_amount",
  "TRANSACTION_EVENT_ID",
  "ACTIVITY_TRANSACTION_ID",
];
writeTable(detailSheet, "A1", detailHeaders, data.detail);
detailSheet.getRange("L2:N5000").format.numberFormat = "#,##0.00";
detailSheet.freezePanes.freezeRows(1);

for (const sheet of [finalSheet, workflowSheet, currencySheet, detailSheet]) {
  sheet.getUsedRange().format.autofitColumns();
}

const check = await workbook.inspect({
  kind: "table",
  range: "最终输出!A1:E18",
  include: "values,formulas",
  tableMaxRows: 20,
  tableMaxCols: 8,
});
console.log(check.ndjson);

const errors = await workbook.inspect({
  kind: "match",
  searchTerm: "#REF!|#DIV/0!|#VALUE!|#NAME\\?|#N/A",
  options: { useRegex: true, maxResults: 50 },
  summary: "final formula error scan",
});
console.log(errors.ndjson);

const preview = await workbook.render({
  sheetName: "最终输出",
  autoCrop: "all",
  scale: 1,
  format: "png",
});
await fs.writeFile(path.join(outDir, "最终输出.png"), new Uint8Array(await preview.arrayBuffer()));

const xlsx = await SpreadsheetFile.exportXlsx(workbook);
const xlsxPath = path.join(outDir, `A1-A3未代扣代缴输出_${data.activity_period}.xlsx`);
await xlsx.save(xlsxPath);
console.log(xlsxPath);

File: build_workbook.py
from pathlib import Path

from openpyxl import Workbook
from openpyxl.styles import Alignment, Font, PatternFill
from openpyxl.utils import get_column_letter


TITLE_FILL = "17365D"
HEADER_FILL = "1F4E78"
LABEL_FILL = "D9EAF7"
REVIEW_FILL = "FFF2CC"


def _write_table(sheet, start_row, start_col, headers, rows):
    for offset, header in enumerate(headers):
        cell = sheet.cell(row=start_row, column=start_col + offset, value=header)
        cell.fill = PatternFill("solid", fgColor=HEADER_FILL)
        cell.font = Font(bold=True, color="FFFFFF")
        cell.alignment = Alignment(wrap_text=True, vertical="top")

    for row_index, row in enumerate(rows, start=start_row + 1):
        for col_offset, header in enumerate(headers):
            value = row.get(header, "")
            cell = sheet.cell(row=row_index, column=start_col + col_offset, value=value)
            cell.alignment = Alignment(wrap_text=True, vertical="top")


def _autosize(sheet, max_width=60):
    for column_cells in sheet.columns:
        column_letter = get_column_letter(column_cells[0].column)
        width = 10
        for cell in column_cells:
            if cell.value is None:
                continue
            text = str(cell.value)
            width = max(width, min(max_width, len(text) + 2))
        sheet.column_dimensions[column_letter].width = width


def _format_amount_columns(sheet, columns, start_row=2):
    for column in columns:
        for row in range(start_row, sheet.max_row + 1):
            sheet[f"{column}{row}"].number_format = "#,##0.00"


def build_workbook(data, output_path):
    output_path = Path(output_path)
    output_path.parent.mkdir(parents=True, exist_ok=True)

    wb = Workbook()
    final_sheet = wb.active
    final_sheet.title = "最终输出"
    detail_sheet = wb.create_sheet("命中明细")

    final_sheet.sheet_view.showGridLines = False
    final_sheet.merge_cells("A1:F1")
    country_name = data.get("country_name", data.get("tax_country", ""))
    final_sheet["A1"] = f"{country_name}税务工作流输出"
    final_sheet["A1"].fill = PatternFill("solid", fgColor=TITLE_FILL)
    final_sheet["A1"].font = Font(bold=True, color="FFFFFF", size=16)

    info_rows = [
        ("算税期间", data.get("activity_period", "")),
        ("算税国家", data.get("tax_country", "")),
        ("源文件", data.get("source_file", "")),
        ("已注册税号国家", ", ".join(data.get("registered_vat_countries", []))),
        ("未注册欧盟国家", ", ".join(data.get("unregistered_eu_countries", []))),
        (
            "期间汇率",
            ", ".join(
                f"{key}:{value if value is not None else '未填'}"
                for key, value in data.get("rates_for_period", data.get("rates_for_month", {})).items()
            ),
        ),
        ("汇率来源", data.get("exchange_rate_source", "")),
    ]
    for row_index, (label, value) in enumerate(info_rows, start=3):
        final_sheet.cell(row=row_index, column=1, value=label).fill = PatternFill("solid", fgColor=LABEL_FILL)
        final_sheet.cell(row=row_index, column=1).font = Font(bold=True)
        final_sheet.cell(row=row_index, column=2, value=value).alignment = Alignment(wrap_text=True)

    summary_start = 10
    summary_headers = ["项目", "名称", "命中行数", "EUR金额总和", "税金EUR", "缺少汇率行数"]
    _write_table(final_sheet, summary_start, 1, summary_headers, data.get("summary", []))
    final_start = summary_start + len(data.get("summary", [])) + 3
    _write_table(final_sheet, final_start, 1, ["项目", "公式", "金额EUR", "缺少汇率行数"], data.get("final", []))
    _format_amount_columns(final_sheet, ["D", "E"], summary_start + 1)
    _format_amount_columns(final_sheet, ["C"], final_start + 1)
    final_sheet.freeze_panes = "A11"

    detail_rows = data.get("detail", [])
    detail_headers = []
    for row in detail_rows:
        for header in row:
            if header not in detail_headers:
                detail_headers.append(header)
    _write_table(detail_sheet, 1, 1, detail_headers, detail_rows)
    if "人工确认提示" in detail_headers:
        review_column = detail_headers.index("人工确认提示") + 1
        review_fill = PatternFill("solid", fgColor=REVIEW_FILL)
        for row_index in range(2, detail_sheet.max_row + 1):
            if detail_sheet.cell(row_index, review_column).value:
                for column_index in range(1, detail_sheet.max_column + 1):
                    detail_sheet.cell(row_index, column_index).fill = review_fill
    detail_sheet.freeze_panes = "A2"

    for sheet in wb.worksheets:
        _autosize(sheet)
        for row in sheet.iter_rows():
            for cell in row:
                cell.alignment = Alignment(wrap_text=True, vertical="top")

    wb.save(output_path)
    return output_path
